File: lib/features/fees/services/fee_payment_analytics.dart
import 'package:viro_team_v2/services/analytics_service.dart';

/// Événements PostHog du parcours cotisations / Stripe (sans PII).
class FeePaymentAnalytics {
  FeePaymentAnalytics({required AnalyticsService analytics})
      : _analytics = analytics;

  final AnalyticsService _analytics;

  static const String startedEvent = 'fee_payment_started';
  static const String submittedEvent = 'fee_payment_submitted';
  static const String cancelledEvent = 'fee_payment_cancelled';
  static const String failedEvent = 'fee_payment_failed';
  static const String connectStartedEvent = 'fee_connect_started';
  static const String connectFailedEvent = 'fee_connect_failed';

  static const String surfaceApp = 'app';
  static const String surfacePortal = 'portal';

  /// Callable checkout OK, avant PaymentSheet / Payment Element.
  void trackStarted({
    required String surface,
    required int amountCents,
    required String currency,
    required int aidCount,
    required bool hasSession,
  }) {
    _analytics.capture(startedEvent, {
      'surface': surface,
      'amount_cents': amountCents,
      'currency': currency.toLowerCase(),
      'aid_count': aidCount,
      'has_session': hasSession,
    });
  }

  /// Confirmation client (Sheet/Element) — le crédit reste côté webhook.
  void trackSubmitted({
    required String surface,
    required int amountCents,
    required String currency,
    required int aidCount,
  }) {
    _analytics.capture(submittedEvent, {
      'surface': surface,
      'amount_cents': amountCents,
      'currency': currency.toLowerCase(),
      'aid_count': aidCount,
      'provider': 'stripe',
    });
  }

  /// Annulation utilisateur.
  void trackCancelled({
    required String surface,
    required String stage,
  }) {
    _analytics.capture(cancelledEvent, {
      'surface': surface,
      'stage': stage,
    });
  }

  /// Échec callable / Stripe / exception (code technique uniquement).
  void trackFailed({
    required String surface,
    required String stage,
    String errorCode = 'unknown',
  }) {
    _analytics.capture(failedEvent, {
      'surface': surface,
      'stage': stage,
      'error_code': errorCode,
    });
  }

  /// Clic onboarding Stripe Connect (portail).
  void trackConnectStarted({required String surface}) {
    _analytics.capture(connectStartedEvent, {
      'surface': surface,
    });
  }

  /// Échec Connect / refresh status.
  void trackConnectFailed({
    required String surface,
    String errorCode = 'unknown',
  }) {
    _analytics.capture(connectFailedEvent, {
      'surface': surface,
      'error_code': errorCode,
    });
  }
}

File: portal/src/lib/fees/feePaymentAnalytics.ts
import { posthog } from "@/lib/posthog";

const STARTED_EVENT = "fee_payment_started";
const SUBMITTED_EVENT = "fee_payment_submitted";
const CANCELLED_EVENT = "fee_payment_cancelled";
const FAILED_EVENT = "fee_payment_failed";
const CONNECT_STARTED_EVENT = "fee_connect_started";
const CONNECT_FAILED_EVENT = "fee_connect_failed";

const SURFACE_PORTAL = "portal";

/** Capture PostHog best-effort (no-op si SDK non chargé / opt-out). */
function capture(
  eventName: string,
  properties: Record<string, string | number | boolean>,
): void {
  try {
    if (!posthog.__loaded) return;
    posthog.capture(eventName, properties);
  } catch {
    // L'analytics ne doit jamais bloquer le paiement.
  }
}

/** Code d’erreur technique sans message utilisateur ni PII. */
export function feePaymentErrorCode(err: unknown): string {
  if (err && typeof err === "object") {
    const withCode = err as { code?: unknown; name?: unknown };
    if (typeof withCode.code === "string" && withCode.code.trim()) {
      return withCode.code;
    }
    if (typeof withCode.name === "string" && withCode.name.trim()) {
      return withCode.name;
    }
  }
  if (err instanceof Error && err.name) return err.name;
  return "unknown";
}

/** Événements PostHog du parcours cotisations / Stripe (sans PII). */
export const FeePaymentAnalytics = {
  trackStarted(params: {
    amountCents: number;
    currency: string;
    aidCount?: number;
    hasSession: boolean;
  }): void {
    capture(STARTED_EVENT, {
      surface: SURFACE_PORTAL,
      amount_cents: params.amountCents,
      currency: params.currency.toLowerCase(),
      aid_count: params.aidCount ?? 0,
      has_session: params.hasSession,
    });
  },

  trackSubmitted(params: {
    amountCents?: number;
    currency?: string;
    aidCount?: number;
  }): void {
    capture(SUBMITTED_EVENT, {
      surface: SURFACE_PORTAL,
      amount_cents: params.amountCents ?? 0,
      currency: (params.currency ?? "eur").toLowerCase(),
      aid_count: params.aidCount ?? 0,
      provider: "stripe",
    });
  },

  trackCancelled(stage: "element" | "sheet" = "element"): void {
    capture(CANCELLED_EVENT, {
      surface: SURFACE_PORTAL,
      stage,
    });
  },

  trackFailed(params: {
    stage: "callable" | "sheet" | "element";
    errorCode?: string;
  }): void {
    capture(FAILED_EVENT, {
      surface: SURFACE_PORTAL,
      stage: params.stage,
      error_code: params.errorCode ?? "unknown",
    });
  },

  trackConnectStarted(): void {
    capture(CONNECT_STARTED_EVENT, {
      surface: SURFACE_PORTAL,
    });
  },

  trackConnectFailed(errorCode = "unknown"): void {
    capture(CONNECT_FAILED_EVENT, {
      surface: SURFACE_PORTAL,
      error_code: errorCode,
    });
  },
};
